lib/pacemaker/pcmk_sched_instances.c

```diff
@@ -601,35 +601,57 @@ free_instance_list(const pe_resource_t *rsc, GList *list)
     }
 }
 
-gboolean
-is_child_compatible(const pe_resource_t *child_rsc, const pe_node_t *local_node,
-                    enum rsc_role_e filter, gboolean current)
+/*!
+ * \internal
+ * \brief Check whether an instance is compatible with a role and node
+ *
+ * \param[in] instance  Clone instance or bundle replica container
+ * \param[in] node      Instance must match this node
+ * \param[in] role      If not RSC_ROLE_UNKNOWN, instance must match this role
+ * \param[in] current   If true, compare instance's original node and role,
+ *                      otherwise compare assigned next node and role
+ *
+ * \return true if \p instance is compatible with \p node and \p role,
+ *         otherwise false
```

Comment:
This is much cleaner.

Nitpicking the existing nomenclature (no action needed right now)... It feels misleading to say the instance is "compatible" with a particular node/role, when we're checking where the instance is either currently running or assigned -- not where the instance *could* run. I haven't looked through the whole call chain to see how much sense this makes in context. This function just doesn't do what I expected from the brief description.

Reply:
Okay, the rest of the call chain is the rest of the PR :)

This makes sense in the context that we're checking whether an instance is compatible with some other resource (in the caller) whose current or assigned state is the given node and role.

Reply:
Good point, I'll go with `pcmk__instance_matches()`